walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,120 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  private ArrayList<TimeRange> getMergedTimes(ArrayList<TimeRange> takenTimes) {
+    Collections.sort(takenTimes, TimeRange.ORDER_BY_START);
+    TimeRange previous = takenTimes.get(0);
+    ArrayList<TimeRange> takenTimesMerged = new ArrayList();
+
+    for (TimeRange time : takenTimes) {
+      if (time.overlaps(previous)) {
+        int start = previous.start();
+        int end = Math.max(time.end(), previous.end());
+        previous = TimeRange.fromStartEnd(start, end, false);
+      } else {
+          takenTimesMerged.add(previous);
+          previous = time;
+        }
+    }
+    takenTimesMerged.add(previous);
+    return takenTimesMerged;
+  }
+
+  private ArrayList<TimeRange> getBetweenTimes(ArrayList<TimeRange> takenTimesMerged) {
+    ArrayList<TimeRange> validTimes = new ArrayList();
+    TimeRange firstValid = takenTimesMerged.get(0);
+    TimeRange lastValid = takenTimesMerged.get(takenTimesMerged.size() - 1);
+    int first = firstValid.start();
+    int last = lastValid.end();
+    validTimes.add(TimeRange.fromStartEnd(0, first ,false));
```

Comment:
While this ends up being fine since you filter out time slots that are too small below, it's kinda weird to potentially add a time slot with a duration of 0. Not something that necessarily needs to be fixed but just a comment.